crates/composition/src/ingest_subgraph/schema_definitions.rs

```diff
@@ -0,0 +1,228 @@
+            #[cfg(test)]
+            key: Cow::Owned(format!("{DEFAULT_FEDERATION_PREFIX}key")),
+        }
+    }
+}
+
+impl<'a> FederationDirectivesMatcher<'a> {
+    pub(crate) fn is_federation_directive(directive: &ast::ConstDirective) -> bool {
+        directive
+            .get_argument("url")
+            .map(|url| match &url.node {
+                ConstValue::String(s) => s.contains("dev/federation/v2"),
+                _ => false,
+            })
+            .unwrap_or_default()
+    }
+
+    /// Matcher for federation directives in a given subgraph. See [FederationDirectivesMatcher] for more docs.        
+    pub(crate) fn new(directive: &'a ast::ConstDirective) -> FederationDirectivesMatcher<'a> {
+        let mut r#as = None;
+        let mut imported: Vec<(&str, &str)> = Vec::new();
+
+        for (arg_name, arg_value) in &directive.arguments {
+            match (arg_name.node.as_str(), &arg_value.node) {
+                ("as", ConstValue::String(value)) => r#as = Some(value.as_str()),
+                ("import", ConstValue::List(imports)) => read_imports(imports, &mut imported),
+                _ => (),
+            }
+        }
+
+        let federation_prefix = r#as
+            .map(|prefix| Cow::Owned(format!("{prefix}__")))
+            .unwrap_or(Cow::Borrowed(DEFAULT_FEDERATION_PREFIX));
+        let final_name = |directive_name: &str| {
+            imported
+                .iter()
+                .find(|(original, _alias)| *original == directive_name)
+                .map(|(_, alias)| Cow::Borrowed(*alias))
+                .unwrap_or_else(|| Cow::Owned(format!("{federation_prefix}{directive_name}")))
+        };
+
+        FederationDirectivesMatcher {
+            shareable: final_name("shareable"),
+            #[cfg(test)]
+            key: final_name("key"),
+        }
+    }
+
+    pub(crate) fn is_shareable(&self, directive_name: &str) -> bool {
+        self.shareable == directive_name
+    }
+
+    #[cfg(test)]
+    pub(crate) fn is_key(&self, directive_name: &str) -> bool {
+        self.key == directive_name
+    }
+}
+
+fn read_imports<'a>(ast_imports: &'a [ConstValue], out: &mut Vec<(&'a str, &'a str)>) {
+    for import in ast_imports {
+        match import {
+            ConstValue::String(import) => {
+                let import = import.trim_start_matches('@');
+                out.push((import, import));
+            }
+            ConstValue::Object(obj) => {
+                if let Some(ConstValue::String(name)) = obj.get("name") {
+                    let alias = obj.get("as").and_then(|value| match value {
+                        ConstValue::String(s) => Some(s),
+                        _ => None,
+                    });
+                    out.push((
+                        name.trim_start_matches('@'),
+                        alias.unwrap_or(name).trim_start_matches('@'),
+                    ));
+                }
+            }
+            _ => (),
+        }
+    }
+}
```

Comment:
I don't have particularly strong opinions but `ConstValue` implements `Deserializer` so we _could_ use serde to derive this

Reply:
Oh I didn't know that. I'll give it a try and push if that works too and that saves us work.